<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fluent genomics with plyranges and tximeta • fluentGenomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/journal/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Fluent genomics with plyranges and tximeta">
<meta property="og:description" content="fluentGenomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fluentGenomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/fluentGenomics.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sa-lee/fluentGenomics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Fluent genomics with plyranges and tximeta</h1>
                        <h4 class="author">Stuart Lee, Michael Lawrence, Michael Love</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sa-lee/fluentGenomics/blob/master/vignettes/fluentGenomics.Rmd"><code>vignettes/fluentGenomics.Rmd</code></a></small>
      <div class="hidden name"><code>fluentGenomics.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      We construct a simple workflow for fluent genomics data analysis using the R/Bioconductor ecosystem. This involves three core steps: <strong>import</strong> the data into an appropriate abstraction, <strong>model</strong> the data with respect to the biological questions of interest, and <strong>integrate</strong> the results with respect to their underlying genomic coordinates. Here we show how to implement these steps to integrate published RNA-seq and ATAC-seq experiments on macrophage cell lines. Using <em>tximeta</em>, we <strong>import</strong> RNA-seq transcript quantifications into an analysis-ready data structure, called the <em>SummarizedExperiment</em>, that contains the ranges of the reference transcripts and metadata on their provenance. Using <em>SummarizedExperiment</em>s to represent the ATAC-seq and RNA-seq data, we <strong>model</strong> differentially accessible (DA) chromatin peaks and differentially expressed (DE) genes with existing Bioconductor packages. Using <em>plyranges</em> we then <strong>integrate</strong> the results to see if there is an enrichment of DA peaks near DE genes by finding overlaps and aggregating over log-fold change thresholds. The combination of these packages and their integration with the Bioconductor ecosystem provide a coherent framework for analysts to iteratively and reproducibly explore their biological data.
    </div>
    
<p>(ref:workflow) An overview of the fluent genomics workflow. First, we <em>import</em> data as a <em>SummarizedExperiment</em> object, which enables interoperability with downstream analysis packages. Then we <em>model</em> our assay data, using the existing Bioconductor packages <em>DESeq2</em> and <em>limma</em>. We take the results of our models for each assay with respect to their genomic coordinates, and <em>integrate</em> them. First, we compute the overlap between the results of each assay, then aggregate over the combined genomic regions, and finally summarize to compare enrichment for differentially expressed genes to non differentially expressed genes. The final output can be used for downstream visualization or further transformation.</p>
<p>(ref:maplot) Visualization of <em>DESeq2</em> results as an “MA plot”. Genes that have an adjusted <em>p-value</em> below 0.01 are colored red.</p>
<p>(ref:boxplot) A boxplot of maximum LFCs for DA peaks for DE genes compared to non-DE genes where genes have at least one DA peak.</p>
<p>(ref:linechart) A line chart displaying how relative enrichment of DA peaks change between DE genes compared to non-DE genes as the absolute DA LFC threshold increases.</p>
<p>(ref:linechart2) A line chart displaying how gene and peak counts change as the absolute DA LFC threshold increases. Lines are colored according to whether they represent a gene that is DE or not. Note the x-axis is on a <span class="math inline">\(\log_{10}\)</span> scale.</p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>In this workflow, we examine a subset of the RNA-seq and ATAC-seq data from <span class="citation">Alasoo et al. (<a href="#ref-alasoo">2018</a>)</span>, a study that involved treatment of macrophage cell lines from a number of human donors with interferon gamma (IFNg), <em>Salmonella</em> infection, or both treatments combined. <span class="citation">Alasoo et al. (<a href="#ref-alasoo">2018</a>)</span> examined gene expression and chromatin accessibility in a subset of 86 successfully differentiated induced pluripotent stem cells (iPSC) lines, and compared baseline and response with respect to chromatin accessibility and gene expression at specific quantitative trait loci (QTL). The authors found that many of the stimulus-specific expression QTL were already detectable as chromatin QTL in naive cells, and further hypothesize about the nature and role of transcription factors implicated in the response to stimulus.</p>
<p>We will perform a much simpler analysis than the one found in <span class="citation">Alasoo et al. (<a href="#ref-alasoo">2018</a>)</span>, using their publicly available RNA-seq and ATAC-seq data (ignoring the genotypes). We will examine the effect of IFNg stimulation on gene expression and chromatin accessibility, and look to see if there is an enrichment of differentially accessible (DA) ATAC-seq peaks in the vicinity of differentially expressed (DE) genes. This is plausible, as the transcriptomic response to IFNg stimulation may be mediated through binding of regulatory proteins to accessible regions, and this binding may increase the accessibility of those regions such that it can be detected by ATAC-seq.</p>
<div class="figure" style="text-align: center">
<img src="workflow.png" alt="(ref:workflow)" width="\textwidth"><p class="caption">
(ref:workflow)
</p>
</div>
<p>Throughout the workflow (Figure @ref(fig:workflow)), we will use existing Bioconductor infrastructure to understand these datasets. In particular, we will emphasize the use of the Bioconductor packages <em>plyranges</em> and <em>tximeta</em>. The <em>plyranges</em> package fluently transforms data tied to genomic ranges using operations like shifting, window construction, overlap detection, etc. It is described by <span class="citation">Lee, Cook, and Lawrence (<a href="#ref-Lee2019">2019</a>)</span> and leverages underlying core Bioconductor infrastructure <span class="citation">(Lawrence et al. <a href="#ref-granges">2013</a>; Huber et al. <a href="#ref-bioc">2015</a>)</span> and the <em>tidyverse</em> design principles <span class="citation">Wickham et al. (<a href="#ref-tidyverse">2019</a>)</span>.</p>
<p>The <em>tximeta</em> package described by <span class="citation">Love et al. (<a href="#ref-Love2019-tximeta">2019</a>)</span> is used to read RNA-seq quantification data into R/Bioconductor, such that the transcript ranges and their provenance are automatically attached to the object containing expression values and differential expression results.</p>
<div id="experimental-data" class="section level2">
<h2 class="hasAnchor">
<a href="#experimental-data" class="anchor"></a>Experimental Data</h2>
<p>The data used in this workflow is available from two packages: the <em>macrophage</em> Bioconductor ExperimentData package and from the workflow package <em>fluentGenomics</em>.</p>
<p>The <em>macrophage</em> package contains RNA-seq quantification from 24 RNA-seq samples, a subset of the RNA-seq samples generated and analyzed by <span class="citation">Alasoo et al. (<a href="#ref-alasoo">2018</a>)</span>. The paired-end reads were quantified using <em>Salmon</em> <span class="citation">(Patro et al. <a href="#ref-salmon">2017</a>)</span>, using the Gencode 29 human reference transcripts <span class="citation">(Frankish, GENCODE-consoritum, and Flicek <a href="#ref-gencode">2018</a>)</span>. For more details on quantification, and the exact code used, consult the vignette of the <a href="http://bioconductor.org/packages/macrophage">macrophage</a> package. The package also contains the <code>Snakemake</code> file that was used to distribute the <em>Salmon</em> quantification jobs on a cluster <span class="citation">(Köster and Rahmann <a href="#ref-snakemake">2012</a>)</span>.</p>
<p>The <em>fluentGenomics</em> package contains functionality to download and generate a cached <em>SummarizedExperiment</em> object from the normalized ATAC-seq data provided by <span class="citation">Alasoo and Gaffney (<a href="#ref-alasooZenodo">2017</a>)</span>. This object contains all 145 ATAC-seq samples across all experimental conditions as analyzed by <span class="citation">Alasoo et al. (<a href="#ref-alasoo">2018</a>)</span>. The data can be also be downloaded directly from the <a href="https://zenodo.org/record/1188300#.XIAhXlNKjOQ">Zenodo</a> deposition.</p>
<p>The following code loads the path to the cached data file, or if it is not present, will create the cache and generate a <em>SummarizedExperiment</em> using the the <em>BiocFileCache</em> package <span class="citation">(Shepherd and Morgan <a href="#ref-bcfilecache">2019</a>)</span>.</p>
<p>We can then read the cached file and assign it to an object called <code>atac</code>.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">fluentGenomics</span>)
<span class="no">atac</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="fu"><a href="../reference/cache-se.html">cache_atac_se</a></span>())</pre></body></html></div>
<p>A precise description of how we obtained this <em>SummarizedExperiment</em> object can be found in section @ref(atac).</p>
</div>
</div>
<div id="se" class="section level1">
<h1 class="hasAnchor">
<a href="#se" class="anchor"></a>Import Data as a <em>SummarizedExperiment</em>
</h1>
<div id="using-tximeta-to-import-rna-seq-quantification-data" class="section level2">
<h2 class="hasAnchor">
<a href="#using-tximeta-to-import-rna-seq-quantification-data" class="anchor"></a>Using <em>tximeta</em> to import RNA-seq quantification data</h2>
<p>First, we specify a directory <code>dir</code>, where the quantification files are stored. You could simply specify this directory with:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">dir</span> <span class="kw">&lt;-</span> <span class="st">"/path/to/quant/files"</span></pre></body></html></div>
<p>where the path is relative to your current R session. However, in this case we have distributed the files in the <em>macrophage</em> package. The relevant directory and associated files can be located using <code>system.file</code>.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">dir</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"macrophage"</span>)</pre></body></html></div>
<p>Information about the experiment is contained in the <code>coldata.csv</code> file. We leverage the <em>dplyr</em> and <em>readr</em> packages (as part of the <em>tidyverse</em>) to read this file into R <span class="citation">(Wickham et al. <a href="#ref-tidyverse">2019</a>)</span>. We will see later that <em>plyranges</em> extends these packages to accommodate genomic ranges.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)</pre></body></html></div>
<pre><code>## 
## Attaching package: 'dplyr'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">readr</span>)
<span class="no">colfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">dir</span>, <span class="st">"coldata.csv"</span>)
<span class="no">coldata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span>(<span class="no">colfile</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(
    <span class="no">names</span>,
    <span class="kw">id</span> <span class="kw">=</span> <span class="no">sample_id</span>,
    <span class="kw">line</span> <span class="kw">=</span> <span class="no">line_id</span>,
    <span class="kw">condition</span> <span class="kw">=</span> <span class="no">condition_name</span>
  ) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(
    <span class="kw">files</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">dir</span>, <span class="st">"quants"</span>, <span class="no">names</span>, <span class="st">"quant.sf.gz"</span>),
    <span class="kw">line</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">line</span>),
    <span class="kw">condition</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">condition</span>), <span class="st">"naive"</span>)
  )</pre></body></html></div>
<pre><code>## Parsed with column specification:
## cols(
##   names = col_character(),
##   sample_id = col_character(),
##   line_id = col_character(),
##   replicate = col_double(),
##   condition_name = col_character(),
##   macrophage_harvest = col_character(),
##   salmonella_date = col_character(),
##   ng_ul_mean = col_double(),
##   rna_extraction = col_character(),
##   rna_submit = col_character(),
##   library_pool = col_character(),
##   chemistry = col_character(),
##   rna_auto = col_double()
## )</code></pre>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">coldata</span></pre></body></html></div>
<pre><code>## # A tibble: 24 x 5
##    names      id     line  condition  files                                     
##    &lt;chr&gt;      &lt;chr&gt;  &lt;fct&gt; &lt;fct&gt;      &lt;chr&gt;                                     
##  1 SAMEA1038… diku_A diku… naive      /usr/local/lib/R/host-site-library/macrop…
##  2 SAMEA1038… diku_B diku… IFNg       /usr/local/lib/R/host-site-library/macrop…
##  3 SAMEA1038… diku_C diku… SL1344     /usr/local/lib/R/host-site-library/macrop…
##  4 SAMEA1038… diku_D diku… IFNg_SL13… /usr/local/lib/R/host-site-library/macrop…
##  5 SAMEA1038… eiwy_A eiwy… naive      /usr/local/lib/R/host-site-library/macrop…
##  6 SAMEA1038… eiwy_B eiwy… IFNg       /usr/local/lib/R/host-site-library/macrop…
##  7 SAMEA1038… eiwy_C eiwy… SL1344     /usr/local/lib/R/host-site-library/macrop…
##  8 SAMEA1038… eiwy_D eiwy… IFNg_SL13… /usr/local/lib/R/host-site-library/macrop…
##  9 SAMEA1038… fikt_A fikt… naive      /usr/local/lib/R/host-site-library/macrop…
## 10 SAMEA1038… fikt_B fikt… IFNg       /usr/local/lib/R/host-site-library/macrop…
## # … with 14 more rows</code></pre>
<p>After we have read the <code>coldata.csv</code> file, we select relevant columns from this table, create a new column called <code>files</code>, and transform the existing <code>line</code> and <code>condition</code> columns into factors. In the case of <code>condition</code>, we specify the “naive” cell line as the reference level. The <code>files</code> column points to the quantifications for each observation – these files have been gzipped, but would typically not have the ‘gz’ ending if used from <em>Salmon</em> directly. One other thing to note is the use of the pipe operator,<code><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">%&gt;%</a></code>, which can be read as “then”, i.e. first read the data, <em>then</em> select columns, <em>then</em> mutate them.</p>
<p>Now we have a table summarizing the experimental design and the locations of the quantifications. The following lines of code do a lot of work for the analyst: importing the RNA-seq quantification (dropping <em>inferential replicates</em> in this case), locating the relevant reference transcriptome, attaching the transcript ranges to the data, and fetching genome information. Inferential replicates are especially useful for performing transcript-level analysis, but here we will use a point estimate for the per-gene counts and perform gene-level analysis.</p>
<p>The result is a <em>SummarizedExperiment</em> object.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SummarizedExperiment</span>))
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tximeta</span>)
<span class="no">se</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximeta/man/tximeta.html">tximeta</a></span>(<span class="no">coldata</span>, <span class="kw">dropInfReps</span><span class="kw">=</span><span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## importing quantifications</code></pre>
<pre><code>## reading in files with read_tsv</code></pre>
<pre><code>## 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 
## found matching linked transcriptome:
## [ GENCODE - Homo sapiens - release 29 ]
## building TxDb with 'GenomicFeatures' package
## Import genomic features from the file as a GRanges object ... OK
## Prepare the 'metadata' data frame ... OK
## Make the TxDb object ...</code></pre>
<pre><code>## Warning in .get_cds_IDX(mcols0$type, mcols0$phase): The "phase" metadata column contains non-NA values for features of type
##   stop_codon. This information was ignored.</code></pre>
<pre><code>## OK
## generating transcript ranges
## fetching genome info for GENCODE</code></pre>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">se</span></pre></body></html></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 205870 24 
## metadata(6): tximetaInfo quantInfo ... txomeInfo txdbInfo
## assays(3): counts abundance length
## rownames(205870): ENST00000456328.2 ENST00000450305.2 ...
##   ENST00000387460.2 ENST00000387461.2
## rowData names(3): tx_id gene_id tx_name
## colnames(24): SAMEA103885102 SAMEA103885347 ... SAMEA103885308
##   SAMEA103884949
## colData names(4): names id line condition</code></pre>
<!-- should we describe the data structure in more detail, or use a figure -->
<p>On a machine with a working internet connection, the above command works without any extra steps, as the <code>tximeta</code> function obtains any necessary metadata via FTP, unless it is already cached locally. The <em>tximeta</em> package can also be used without an internet connection, in this case the linked transcriptome can be created directly from a <em>Salmon</em> index and gtf.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/tximeta/man/linkedTxome.html">makeLinkedTxome</a></span>(
  <span class="kw">indexDir</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">dir</span>, <span class="st">"gencode.v29_salmon_0.12.0"</span>),
  <span class="kw">source</span><span class="kw">=</span><span class="st">"Gencode"</span>,
  <span class="kw">organism</span><span class="kw">=</span><span class="st">"Homo sapiens"</span>,
  <span class="kw">release</span><span class="kw">=</span><span class="st">"29"</span>,
  <span class="kw">genome</span><span class="kw">=</span><span class="st">"GRCh38"</span>,
  <span class="kw">fasta</span><span class="kw">=</span><span class="st">"ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_29/gencode.v29.transcripts.fa.gz"</span>,
  <span class="kw">gtf</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">dir</span>, <span class="st">"gencode.v29.annotation.gtf.gz"</span>), <span class="co"># local version</span>
  <span class="kw">write</span><span class="kw">=</span><span class="fl">FALSE</span>
)</pre></body></html></div>
<p>Because <em>tximeta</em> knows the correct reference transcriptome, we can ask <em>tximeta</em> to summarize the transcript-level data to the gene level using the methods of <span class="citation">Soneson, Love, and Robinson (<a href="#ref-Soneson2015">2015</a>)</span>.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">gse</span> <span class="kw">&lt;-</span> <span class="fu">summarizeToGene</span>(<span class="no">se</span>)</pre></body></html></div>
<pre><code>## loading existing TxDb created: 2020-05-22 03:02:40</code></pre>
<pre><code>## Loading required package: GenomicFeatures</code></pre>
<pre><code>## Loading required package: AnnotationDbi</code></pre>
<pre><code>## 
## Attaching package: 'AnnotationDbi'</code></pre>
<pre><code>## The following object is masked from 'package:dplyr':
## 
##     select</code></pre>
<pre><code>## obtaining transcript-to-gene mapping from database</code></pre>
<pre><code>## generating gene ranges</code></pre>
<pre><code>## summarizing abundance</code></pre>
<pre><code>## summarizing counts</code></pre>
<pre><code>## summarizing length</code></pre>
<p>One final note is that the <code>start</code> of positive strand genes and the <code>end</code> of negative strand genes is now dictated by the genomic extent of the isoforms of the gene (so the <code>start</code> and <code>end</code> of the reduced <em>GRanges</em>). Another alternative would be to either operate on transcript abundance, and perform differential analysis on transcript (and so avoid defining the TSS of a set of isoforms), or to use gene-level summarized expression but to pick the most representative TSS based on isoform expression.</p>
</div>
<div id="atac" class="section level2">
<h2 class="hasAnchor">
<a href="#atac" class="anchor"></a>Importing ATAC-seq data as a <em>SummarizedExperiment</em> object</h2>
<p>The <em>SummarizedExperiment</em> object containing ATAC-seq peaks can be created from the following tab-delimited files from <span class="citation">Alasoo and Gaffney (<a href="#ref-alasooZenodo">2017</a>)</span>:</p>
<ul>
<li>The sample metadata: <code>ATAC_sample_metadata.txt.gz</code> (&lt;1M)</li>
<li>The matrix of normalized read counts: <code>ATAC_cqn_matrix.txt.gz</code> (109M)</li>
<li>The annotated peaks: <code>ATAC_peak_metadata.txt.gz</code> (5.6M)</li>
</ul>
<p>To begin, we read in the sample metadata, following similar steps to those we used to generate the <code>coldata</code> table for the RNA-seq experiment:</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="no">atac_coldata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(<span class="st">"ATAC_sample_metadata.txt.gz"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(
    <span class="no">sample_id</span>,
    <span class="no">donor</span>,
    <span class="kw">condition</span> <span class="kw">=</span> <span class="no">condition_name</span>
  ) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">condition</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">condition</span>), <span class="st">"naive"</span>))</pre></body></html></div>
<p>The ATAC-seq counts have already been normalized with <em>cqn</em> <span class="citation">(Hansen, Irizarry, and Wu <a href="#ref-Hansen2012">2012</a>)</span> and log2 transformed. Loading the <em>cqn</em>-normalized matrix of log2 transformed read counts takes ~30 seconds and loads an object of ~370 Mb. We set the column names so that the first column contains the rownames of the matrix, and the remaining columns are the sample identities from the <code>atac_coldata</code> object.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">atac_mat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(<span class="st">"ATAC_cqn_matrix.txt.gz"</span>,
                     <span class="kw">skip</span> <span class="kw">=</span> <span class="fl">1</span>,
                     <span class="kw">col_names</span> <span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rownames"</span>, <span class="no">atac_coldata</span><span class="kw">[[</span><span class="st">"sample_id"</span>]]))
<span class="no">rownames</span> <span class="kw">&lt;-</span> <span class="no">atac_mat</span><span class="kw">[[</span><span class="st">"rownames"</span>]]
<span class="no">atac_mat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">atac_mat</span>[,-<span class="fl">1</span>])
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">atac_mat</span>) <span class="kw">&lt;-</span> <span class="no">rownames</span></pre></body></html></div>
<p>We read in the peak metadata (locations in the genome), and convert it to a <em>GRanges</em> object. The <code><a href="https://rdrr.io/pkg/plyranges/man/ranges-construct.html">as_granges()</a></code> function automatically converts the <em>data.frame</em> into a <em>GRanges</em> object. From that result, we extract the peak_id column and set the genome information to the build “GRCh38”. We know this from the <a href="https://zenodo.org/record/1188300#.XJOFSlNKiL5">Zenodo entry</a>.</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">plyranges</span>)
<span class="no">peaks_df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(<span class="st">"ATAC_peak_metadata.txt.gz"</span>,
                     <span class="kw">col_types</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"cidciicdc"</span>)
)

<span class="no">peaks_gr</span> <span class="kw">&lt;-</span> <span class="no">peaks_df</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-construct.html">as_granges</a></span>(<span class="kw">seqnames</span> <span class="kw">=</span> <span class="no">chr</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">select</a></span>(<span class="kw">peak_id</span><span class="kw">=</span><span class="no">gene_id</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-info.html">set_genome_info</a></span>(<span class="kw">genome</span> <span class="kw">=</span> <span class="st">"GRCh38"</span>)</pre></body></html></div>
<p>Finally, we construct a <em>SummarizedExperiment</em> object. We place the matrix into the assays slot as a named list, the annotated peaks into the row-wise ranges slot, and the sample metadata into the column-wise data slot:</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="no">atac</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment</a></span>(<span class="kw">assays</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">cqndata</span><span class="kw">=</span><span class="no">atac_mat</span>),
                             <span class="kw">rowRanges</span><span class="kw">=</span><span class="no">peaks_gr</span>,
                             <span class="kw">colData</span><span class="kw">=</span><span class="no">atac_coldata</span>)</pre></body></html></div>
</div>
</div>
<div id="model-assays" class="section level1">
<h1 class="hasAnchor">
<a href="#model-assays" class="anchor"></a>Model assays</h1>
<div id="rna-seq-differential-gene-expression-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#rna-seq-differential-gene-expression-analysis" class="anchor"></a>RNA-seq differential gene expression analysis</h2>
<p>We can easily run a differential expression analysis with <em>DESeq2</em> using the following code chunks <span class="citation">(Love, Huber, and Anders <a href="#ref-Love2014">2014</a>)</span>. The design formula indicates that we want to control for the donor baselines (<code>line</code>) and test for differences in gene expression on the condition. For a more comprehensive discussion of DE workflows in Bioconductor see <span class="citation">Love et al. (<a href="#ref-Love2016-f1000">2016</a>)</span> and <span class="citation">Law et al. (<a href="#ref-Law2018-f1000">2018</a>)</span>.</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DESeq2</span>)
<span class="no">dds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSet</a></span>(<span class="no">gse</span>, ~<span class="no">line</span> + <span class="no">condition</span>)</pre></body></html></div>
<pre><code>## using counts and average transcript lengths from tximeta</code></pre>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="co"># filter out lowly expressed genes</span>
<span class="co"># at least 10 counts in at least 6 samples</span>
<span class="no">keep</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/counts.html">counts</a></span>(<span class="no">dds</span>) <span class="kw">&gt;=</span> <span class="fl">10</span>) <span class="kw">&gt;=</span> <span class="fl">6</span>
<span class="no">dds</span> <span class="kw">&lt;-</span> <span class="no">dds</span>[<span class="no">keep</span>,]</pre></body></html></div>
<p>The model is fit with the following line of code:</p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="no">dds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span>(<span class="no">dds</span>)</pre></body></html></div>
<pre><code>## estimating size factors</code></pre>
<pre><code>## using 'avgTxLength' from assays(dds), correcting for library size</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<p>Below we set the contrast on the condition variable, indicating we are estimating the <span class="math inline">\(\log_2\)</span> fold change (LFC) of IFNg stimulated cell lines against naive cell lines. We are interested in LFC greater than 1 at a nominal false discovery rate (FDR) of 1%.</p>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="no">res</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span>(<span class="no">dds</span>,
               <span class="kw">contrast</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"condition"</span>,<span class="st">"IFNg"</span>,<span class="st">"naive"</span>),
               <span class="kw">lfcThreshold</span><span class="kw">=</span><span class="fl">1</span>, <span class="kw">alpha</span><span class="kw">=</span><span class="fl">0.01</span>)</pre></body></html></div>
<p>To see the results of the expression analysis, we can generate a summary table and an MA plot:</p>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/summary.html">summary</a></span>(<span class="no">res</span>)</pre></body></html></div>
<pre><code>## 
## out of 17806 with nonzero total read count
## adjusted p-value &lt; 0.01
## LFC &gt; 1.00 (up)    : 502, 2.8%
## LFC &lt; -1.00 (down) : 247, 1.4%
## outliers [1]       : 0, 0%
## low counts [2]     : 0, 0%
## (mean count &lt; 3)
## [1] see 'cooksCutoff' argument of ?results
## [2] see 'independentFiltering' argument of ?results</code></pre>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="kw pkg">DESeq2</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/plotMA.html">plotMA</a></span>(<span class="no">res</span>, <span class="kw">ylim</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(-<span class="fl">10</span>,<span class="fl">10</span>))</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="fluentGenomics_files/figure-html/ma-plot-1.png" alt="(ref:maplot)" width="700"><p class="caption">
(ref:maplot)
</p>
</div>
<p>We now output the results as a <em>GRanges</em> object, and due to the conventions of <em>plyranges</em>, we construct a new column called <code>gene_id</code> from the row names of the results. Each row now contains the genomic region (<code>seqnames</code>, <code>start</code>, <code>end</code>, <code>strand</code>) along with corresponding metadata columns (the <code>gene_id</code> and the results of the test). Note that <em>tximeta</em> has correctly identified the reference genome as “hg38”, and this has also been added to the <em>GRanges</em> along the results columns. This kind of book-keeping is vital once overlap operations are performed to ensure that <em>plyranges</em> is not comparing across incompatible genomes.</p>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">plyranges</span>))
<span class="no">de_genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span>(<span class="no">dds</span>,
                    <span class="kw">contrast</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"condition"</span>,<span class="st">"IFNg"</span>,<span class="st">"naive"</span>),
                    <span class="kw">lfcThreshold</span><span class="kw">=</span><span class="fl">1</span>,
                    <span class="kw">format</span><span class="kw">=</span><span class="st">"GRanges"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-names.html">names_to_column</a></span>(<span class="st">"gene_id"</span>)
<span class="no">de_genes</span></pre></body></html></div>
<pre><code>## GRanges object with 17806 ranges and 7 metadata columns:
##           seqnames              ranges strand |            gene_id   baseMean
##              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;character&gt;  &lt;numeric&gt;
##       [1]     chrX 100627109-100639991      - | ENSG00000000003.14    171.571
##       [2]    chr20   50934867-50958555      - | ENSG00000000419.12    967.751
##       [3]     chr1 169849631-169894267      - | ENSG00000000457.13    682.433
##       [4]     chr1 169662007-169854080      + | ENSG00000000460.16    262.963
##       [5]     chr1   27612064-27635277      - | ENSG00000000938.12   2660.102
##       ...      ...                 ...    ... .                ...        ...
##   [17802]    chr10   84167228-84172093      - |  ENSG00000285972.1   10.04746
##   [17803]     chr6   63572012-63583587      + |  ENSG00000285976.1 4586.34617
##   [17804]    chr16   57177349-57181390      + |  ENSG00000285979.1   14.29653
##   [17805]     chr8 103398658-103501895      - |  ENSG00000285982.1   27.76296
##   [17806]    chr10   12563151-12567351      + |  ENSG00000285994.1    6.60409
##           log2FoldChange     lfcSE      stat    pvalue      padj
##                &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##       [1]     -0.2822450 0.3005710   0.00000 1.0000000  1.000000
##       [2]      0.0391223 0.0859708   0.00000 1.0000000  1.000000
##       [3]      1.2846179 0.1969067   1.44545 0.1483329  1.000000
##       [4]     -1.4718762 0.2186916  -2.15772 0.0309493  0.409728
##       [5]      0.6754781 0.2360530   0.00000 1.0000000  1.000000
##       ...            ...       ...       ...       ...       ...
##   [17802]      0.5484518  0.444319         0         1         1
##   [17803]     -0.0339296  0.188005         0         1         1
##   [17804]      0.3123477  0.522700         0         1         1
##   [17805]      0.9945187  1.582373         0         1         1
##   [17806]      0.2539975  0.595751         0         1         1
##   -------
##   seqinfo: 25 sequences (1 circular) from hg38 genome</code></pre>
<p>From this, we can restrict the results to those that meet our FDR threshold and select (and rename) the metadata columns we’re interested in:</p>
<div class="sourceCode" id="cb53"><html><body><pre class="r"><span class="no">de_genes</span> <span class="kw">&lt;-</span> <span class="no">de_genes</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">filter</a></span>(<span class="no">padj</span> <span class="kw">&lt;</span> <span class="fl">0.01</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">select</a></span>(<span class="no">gene_id</span>, <span class="kw">de_log2FC</span> <span class="kw">=</span> <span class="no">log2FoldChange</span>, <span class="kw">de_padj</span> <span class="kw">=</span> <span class="no">padj</span>)</pre></body></html></div>
<p>We now wish to extract genes for which there is evidence that the LFC is <em>not</em> large. We perform this test by specifying an LFC threshold and an alternative hypothesis (<code>altHypothesis</code>) that the LFC is less than the threshold in absolute value. To visualize the result of this test, you can run <code>results</code> without <code>format="GRanges"</code>, and pass this object to <code>plotMA</code> as before. We label these genes as <code>other_genes</code> and later as “non-DE genes”, for comparison with our <code>de_genes</code> set.</p>
<div class="sourceCode" id="cb54"><html><body><pre class="r"><span class="no">other_genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span>(<span class="no">dds</span>,
                       <span class="kw">contrast</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"condition"</span>,<span class="st">"IFNg"</span>,<span class="st">"naive"</span>),
                       <span class="kw">lfcThreshold</span><span class="kw">=</span><span class="fl">1</span>,
                       <span class="kw">altHypothesis</span><span class="kw">=</span><span class="st">"lessAbs"</span>,
                       <span class="kw">format</span><span class="kw">=</span><span class="st">"GRanges"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">filter</a></span>(<span class="no">padj</span> <span class="kw">&lt;</span> <span class="fl">0.01</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-names.html">names_to_column</a></span>(<span class="st">"gene_id"</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">gene_id</span>,
                <span class="kw">de_log2FC</span> <span class="kw">=</span> <span class="no">log2FoldChange</span>,
                <span class="kw">de_padj</span> <span class="kw">=</span> <span class="no">padj</span>)</pre></body></html></div>
</div>
<div id="atac-seq-peak-differential-abundance-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#atac-seq-peak-differential-abundance-analysis" class="anchor"></a>ATAC-seq peak differential abundance analysis</h2>
<p>The following section describes the process we have used for generating a <em>GRanges</em> object of differential peaks from the ATAC-seq data in <span class="citation">Alasoo et al. (<a href="#ref-alasoo">2018</a>)</span>.</p>
<p>The code chunks for the remainder of this section are not run.</p>
<!-- We can check the standard deviation over mean plot, to assess for any -->
<!-- systematic trends: -->
<!-- ```{r, eval = FALSE} -->
<!-- library(ggplot2) -->
<!-- mv <-
data.frame(rmu = rowMeans(assay(atac)), -->
<!-- rvar = rowVars(assay(atac)))
-->
<!-- ggplot(data =mv , aes(x = rmu, y = sqrt(rvar))) + -->
<!-- geom_hex()
-->
<!-- ``` -->
<p>For assessing differential accessibility, we run <em>limma</em> <span class="citation">(Smyth <a href="#ref-Smyth2004">2004</a>)</span>, and generate the a summary of LFCs and adjusted p-values for the peaks:</p>
<div class="sourceCode" id="cb55"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">limma</span>)
<span class="no">design</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>(~<span class="no">donor</span> + <span class="no">condition</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span>(<span class="no">atac</span>))
<span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span>(<span class="no">atac</span>), <span class="no">design</span>)
<span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span>(<span class="no">fit</span>)
<span class="no">idx</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">fit</span>$<span class="no">coefficients</span>) <span class="kw">==</span> <span class="st">"conditionIFNg"</span>)
<span class="no">tt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span>(<span class="no">fit</span>, <span class="kw">coef</span><span class="kw">=</span><span class="no">idx</span>, <span class="kw">sort.by</span><span class="kw">=</span><span class="st">"none"</span>, <span class="kw">n</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">atac</span>))</pre></body></html></div>
<p>We now take the <code>rowRanges</code> of the <em>SummarizedExperiment</em> and attach the LFCs and adjusted p-values from <em>limma</em>, so that we can consider the overlap with differential expression. Note that we set the genome build to “hg38” and restyle the chromosome information to use the “UCSC” style (e.g. “chr1”, “chr2”, etc.). Again, we know the genome build from the Zenodo entry for the ATAC-seq data.</p>
<div class="sourceCode" id="cb56"><html><body><pre class="r"><span class="no">atac_peaks</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">rowRanges</a></span>(<span class="no">atac</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-names.html">remove_names</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">mutate</a></span>(
    <span class="kw">da_log2FC</span> <span class="kw">=</span> <span class="no">tt</span>$<span class="no">logFC</span>,
    <span class="kw">da_padj</span> <span class="kw">=</span> <span class="no">tt</span>$<span class="no">adj.P.Val</span>
  ) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-info.html">set_genome_info</a></span>(<span class="kw">genome</span> <span class="kw">=</span> <span class="st">"hg38"</span>)

<span class="fu">seqlevelsStyle</span>(<span class="no">atac_peaks</span>) <span class="kw">&lt;-</span> <span class="st">"UCSC"</span></pre></body></html></div>
<p>The final <em>GRanges</em> object containing the DA peaks is included in the workflow package and can be loaded as follows:</p>
<div class="sourceCode" id="cb57"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">fluentGenomics</span>)
<span class="no">peaks</span></pre></body></html></div>
<pre><code>## GRanges object with 296220 ranges and 3 metadata columns:
##            seqnames              ranges strand |          peak_id da_log2FC
##               &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;character&gt; &lt;numeric&gt;
##        [1]     chr1          9979-10668      * |      ATAC_peak_1  0.266185
##        [2]     chr1         10939-11473      * |      ATAC_peak_2  0.322177
##        [3]     chr1         15505-15729      * |      ATAC_peak_3 -0.574160
##        [4]     chr1         21148-21481      * |      ATAC_peak_4 -1.147066
##        [5]     chr1         21864-22067      * |      ATAC_peak_5 -0.896143
##        ...      ...                 ...    ... .              ...       ...
##   [296216]     chrX 155896572-155896835      * | ATAC_peak_296216 -0.834629
##   [296217]     chrX 155958507-155958646      * | ATAC_peak_296217 -0.147537
##   [296218]     chrX 156016760-156016975      * | ATAC_peak_296218 -0.609732
##   [296219]     chrX 156028551-156029422      * | ATAC_peak_296219 -0.347678
##   [296220]     chrX 156030135-156030785      * | ATAC_peak_296220  0.492442
##                da_padj
##              &lt;numeric&gt;
##        [1] 9.10673e-05
##        [2] 2.03435e-05
##        [3] 3.41708e-08
##        [4] 8.22299e-26
##        [5] 4.79453e-11
##        ...         ...
##   [296216] 1.33546e-11
##   [296217] 3.13015e-01
##   [296218] 3.62339e-09
##   [296219] 6.94823e-06
##   [296220] 7.07664e-13
##   -------
##   seqinfo: 23 sequences from hg38 genome; no seqlengths</code></pre>
</div>
</div>
<div id="integrate-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#integrate-ranges" class="anchor"></a>Integrate ranges</h1>
<div id="finding-overlaps-with-plyranges" class="section level2">
<h2 class="hasAnchor">
<a href="#finding-overlaps-with-plyranges" class="anchor"></a>Finding overlaps with <em>plyranges</em>
</h2>
<p>We have already used <em>plyranges</em> a number of times above, to <code>filter</code>, <code>mutate</code>, and <code>select</code> on <em>GRanges</em> objects, as well as ensuring the correct genome annotation and style has been used. The <em>plyranges</em> package provides a grammar for performing transformations of genomic data <span class="citation">(Lee, Cook, and Lawrence <a href="#ref-Lee2019">2019</a>)</span>. Computations resulting from compositions of <em>plyranges</em> “verbs” are performed using underlying, highly optimized range operations in the <em>GenomicRanges</em> package <span class="citation">(Lawrence et al. <a href="#ref-granges">2013</a>)</span>.</p>
<p>For the overlap analysis, we filter the annotated peaks to have a nominal FDR bound of 1%.</p>
<div class="sourceCode" id="cb59"><html><body><pre class="r"><span class="no">da_peaks</span> <span class="kw">&lt;-</span> <span class="no">peaks</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">filter</a></span>(<span class="no">da_padj</span> <span class="kw">&lt;</span> <span class="fl">0.01</span>)</pre></body></html></div>
<p>We now have <em>GRanges</em> objects that contain DE genes, genes without strong signal of DE, and DA peaks. We are ready to answer the question: is there an enrichment of DA ATAC-seq peaks in the vicinity of DE genes compared to genes without sufficient DE signal?</p>
</div>
<div id="down-sampling-non-differentially-expressed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#down-sampling-non-differentially-expressed-genes" class="anchor"></a>Down sampling non-differentially expressed genes</h2>
<p>As <em>plyranges</em> is built on top of <em>dplyr</em>, it implements methods for many of its verbs for <em>GRanges</em> objects. Here we can use <code>slice</code> to randomly sample the rows of the <code>other_genes</code>. The <code>sample.int</code> function will generate random samples of size equal to the number of DE-genes from the number of rows in <code>other_genes</code>:</p>
<div class="sourceCode" id="cb60"><html><body><pre class="r"><span class="no">size</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">de_genes</span>)
<span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">slice</a></span>(<span class="no">other_genes</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html">n</a></span>(), <span class="no">size</span>))</pre></body></html></div>
<pre><code>## GRanges object with 749 ranges and 3 metadata columns:
##         seqnames            ranges strand |            gene_id  de_log2FC
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;character&gt;  &lt;numeric&gt;
##     [1]    chr16 66802875-66873256      - | ENSG00000159593.14  -0.415055
##     [2]    chr19 52797409-52857600      - | ENSG00000198538.10   0.175975
##     [3]    chr15 74843730-74873365      - | ENSG00000140497.16  -0.601723
##     [4]     chr6 31879759-31897687      - | ENSG00000204371.11  -0.203302
##     [5]    chr17 36948875-37056871      + |  ENSG00000275700.4   0.349455
##     ...      ...               ...    ... .                ...        ...
##   [745]     chr8 73972437-73982783      + | ENSG00000175606.10  0.2752368
##   [746]    chr14 77457870-77469474      + |  ENSG00000100591.7  0.0777779
##   [747]    chr19   1905214-1926013      + |  ENSG00000227500.9  0.0213695
##   [748]    chr20 63861212-63864293      - |  ENSG00000268858.2  0.1773121
##   [749]     chr7 80742538-80922359      - | ENSG00000075223.13 -0.2251939
##             de_padj
##           &lt;numeric&gt;
##     [1] 4.06865e-06
##     [2] 5.18734e-11
##     [3] 8.78434e-03
##     [4] 1.64183e-04
##     [5] 2.44019e-09
##     ...         ...
##   [745] 9.62869e-04
##   [746] 1.39809e-14
##   [747] 1.15319e-19
##   [748] 1.68267e-04
##   [749] 9.38845e-03
##   -------
##   seqinfo: 25 sequences (1 circular) from hg38 genome</code></pre>
<p>We can repeat this many times to create many samples via <code>replicate</code>. By replicating the sub-sampling multiple times, we minimize the variance on the enrichment statistics induced by the sampling process.</p>
<div class="sourceCode" id="cb62"><html><body><pre class="r"><span class="co"># set a seed for the results</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">2019</span>-<span class="fl">08</span>-<span class="fl">02</span>)
<span class="no">boot_genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">replicate</a></span>(<span class="fl">10</span>,
                        <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">slice</a></span>(<span class="no">other_genes</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html">n</a></span>(), <span class="no">size</span>)),
                        <span class="kw">simplify</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p>This creates a list of <em>GRanges</em> objects as a list, and we can bind these together using the <code>bind_ranges</code> function. This function creates a new column called “resample” on the result that identifies each of the input <em>GRanges</em> objects:</p>
<div class="sourceCode" id="cb63"><html><body><pre class="r"><span class="no">boot_genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-bind.html">bind_ranges</a></span>(<span class="no">boot_genes</span>, <span class="kw">.id</span> <span class="kw">=</span> <span class="st">"resample"</span>)</pre></body></html></div>
<p>Similarly, we can then combine the <code>boot_genes</code> <em>GRanges</em>, with the DE <em>GRanges</em> object. As the resample column was not present on the DE <em>GRanges</em> object, this is given a missing value which we recode to a 0 using <code><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">mutate()</a></code></p>
<div class="sourceCode" id="cb64"><html><body><pre class="r"><span class="no">all_genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-bind.html">bind_ranges</a></span>(
  <span class="kw">de</span><span class="kw">=</span><span class="no">de_genes</span>,
  <span class="kw">not_de</span> <span class="kw">=</span> <span class="no">boot_genes</span>,
  <span class="kw">.id</span><span class="kw">=</span><span class="st">"origin"</span>
) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">mutate</a></span>(
    <span class="kw">origin</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">origin</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"not_de"</span>, <span class="st">"de"</span>)),
    <span class="kw">resample</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">resample</span>), <span class="fl">0L</span>, <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>(<span class="no">resample</span>))
  )
<span class="no">all_genes</span></pre></body></html></div>
<pre><code>## GRanges object with 8239 ranges and 5 metadata columns:
##          seqnames              ranges strand |            gene_id de_log2FC
##             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;character&gt; &lt;numeric&gt;
##      [1]     chr1 196651878-196747504      + | ENSG00000000971.15   4.98711
##      [2]     chr6   46129993-46146699      + |  ENSG00000001561.6   1.92722
##      [3]     chr4   17577192-17607972      + | ENSG00000002549.12   2.93373
##      [4]     chr7 150800403-150805120      + |  ENSG00000002933.8   3.16722
##      [5]     chr4   15778275-15853230      + | ENSG00000004468.12   5.40894
##      ...      ...                 ...    ... .                ...       ...
##   [8235]    chr17   43527844-43579620      - | ENSG00000175832.12 -0.240918
##   [8236]    chr17   18260534-18266552      + | ENSG00000177427.12 -0.166059
##   [8237]    chr20   63895182-63936031      + | ENSG00000101152.10  0.250539
##   [8238]     chr1   39081316-39487177      + | ENSG00000127603.25 -0.385054
##   [8239]     chr8   41577187-41625001      + | ENSG00000158669.11  0.155922
##              de_padj  resample   origin
##            &lt;numeric&gt; &lt;integer&gt; &lt;factor&gt;
##      [1] 1.37057e-13         0       de
##      [2] 3.17478e-05         0       de
##      [3] 2.01310e-11         0       de
##      [4] 1.07360e-08         0       de
##      [5] 4.82905e-18         0       de
##      ...         ...       ...      ...
##   [8235] 9.91611e-03        10   not_de
##   [8236] 9.12051e-05        10   not_de
##   [8237] 1.74085e-09        10   not_de
##   [8238] 2.65539e-03        10   not_de
##   [8239] 2.96375e-17        10   not_de
##   -------
##   seqinfo: 25 sequences (1 circular) from hg38 genome</code></pre>
</div>
<div id="expanding-genomic-coordinates-around-the-transcription-start-site" class="section level2">
<h2 class="hasAnchor">
<a href="#expanding-genomic-coordinates-around-the-transcription-start-site" class="anchor"></a>Expanding genomic coordinates around the transcription start site</h2>
<p>Now we would like to modify our gene ranges so they contain the 10 kilobases on either side of their transcription start site (TSS). There are many ways one could do this, but we prefer an approach via the anchoring methods in <em>plyranges</em>. Because there is a mutual dependence between the start, end, width, and strand of a <em>GRanges</em> object, we define anchors to fix one of <code>start</code> and <code>end</code>, while modifying the <code>width</code>. As an example, to extract just the TSS, we can anchor by the 5’ end of the range and modify the width of the range to equal 1.</p>
<div class="sourceCode" id="cb66"><html><body><pre class="r"><span class="no">all_genes</span> <span class="kw">&lt;-</span> <span class="no">all_genes</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-anchor.html">anchor_5p</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">mutate</a></span>(<span class="kw">width</span> <span class="kw">=</span> <span class="fl">1</span>)</pre></body></html></div>
<p>Anchoring by the 5’ end of a range will fix the <code>end</code> of negatively stranded ranges, and fix the <code>start</code> of positively stranded ranges.</p>
<p>We can then repeat the same pattern but this time using <code><a href="https://rdrr.io/pkg/plyranges/man/ranges-anchor.html">anchor_center()</a></code> to tell <em>plyranges</em> that we are making the TSS the midpoint of a range that has total width of 20kb, or 10kb both upstream and downstream of the TSS.</p>
<div class="sourceCode" id="cb67"><html><body><pre class="r"><span class="no">all_genes</span> <span class="kw">&lt;-</span> <span class="no">all_genes</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-anchor.html">anchor_center</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">mutate</a></span>(<span class="kw">width</span><span class="kw">=</span><span class="fl">2</span>*<span class="fl">1e4</span>)</pre></body></html></div>
</div>
<div id="use-overlap-joins-to-find-relative-enrichment" class="section level2">
<h2 class="hasAnchor">
<a href="#use-overlap-joins-to-find-relative-enrichment" class="anchor"></a>Use overlap joins to find relative enrichment</h2>
<p>We are now ready to compute overlaps between RNA-seq genes (our DE set and bootstrap sets) and the ATAC-seq peaks. In <em>plyranges</em>, overlaps are defined as joins between two <em>GRanges</em> objects: a <em>left</em> and a <em>right</em> <em>GRanges</em> object. In an overlap join, a match is any range on the <em>left</em> <em>GRanges</em> that is overlapped by the <em>right</em> <em>GRanges</em>. One powerful aspect of the overlap joins is that the result maintains all (metadata) columns from each of the <em>left</em> and <em>right</em> ranges which makes downstream summaries easy to compute.</p>
<p>To combine the DE genes with the DA peaks, we perform a left overlap join. This returns to us the <code>all_genes</code> ranges (potentially with duplication), but with the metadata columns from those overlapping DA peaks. For any gene that has no overlaps, the DA peak columns will have <code>NA</code>’s.</p>
<div class="sourceCode" id="cb68"><html><body><pre class="r"><span class="no">genes_olap_peaks</span> <span class="kw">&lt;-</span> <span class="no">all_genes</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_left</a></span>(<span class="no">da_peaks</span>)
<span class="no">genes_olap_peaks</span></pre></body></html></div>
<pre><code>## GRanges object with 27766 ranges and 8 metadata columns:
##           seqnames              ranges strand |            gene_id de_log2FC
##              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;character&gt; &lt;numeric&gt;
##       [1]     chr1 196641878-196661877      + | ENSG00000000971.15   4.98711
##       [2]     chr6   46119993-46139992      + |  ENSG00000001561.6   1.92722
##       [3]     chr4   17567192-17587191      + | ENSG00000002549.12   2.93373
##       [4]     chr4   17567192-17587191      + | ENSG00000002549.12   2.93373
##       [5]     chr4   17567192-17587191      + | ENSG00000002549.12   2.93373
##       ...      ...                 ...    ... .                ...       ...
##   [27762]     chr1   39071316-39091315      + | ENSG00000127603.25 -0.385054
##   [27763]     chr1   39071316-39091315      + | ENSG00000127603.25 -0.385054
##   [27764]     chr8   41567187-41587186      + | ENSG00000158669.11  0.155922
##   [27765]     chr8   41567187-41587186      + | ENSG00000158669.11  0.155922
##   [27766]     chr8   41567187-41587186      + | ENSG00000158669.11  0.155922
##               de_padj  resample   origin          peak_id da_log2FC     da_padj
##             &lt;numeric&gt; &lt;integer&gt; &lt;factor&gt;      &lt;character&gt; &lt;numeric&gt;   &lt;numeric&gt;
##       [1] 1.37057e-13         0       de  ATAC_peak_21236 -0.546582 1.15274e-04
##       [2] 3.17478e-05         0       de ATAC_peak_231183  1.453297 9.73225e-17
##       [3] 2.01310e-11         0       de ATAC_peak_193578  0.222371 3.00939e-11
##       [4] 2.01310e-11         0       de ATAC_peak_193579 -0.281615 7.99889e-05
##       [5] 2.01310e-11         0       de ATAC_peak_193580  0.673705 7.60043e-15
##       ...         ...       ...      ...              ...       ...         ...
##   [27762] 2.65539e-03        10   not_de   ATAC_peak_5357 -1.058236 3.69052e-16
##   [27763] 2.65539e-03        10   not_de   ATAC_peak_5358 -1.314112 6.44280e-26
##   [27764] 2.96375e-17        10   not_de ATAC_peak_263396 -0.904080 8.19577e-13
##   [27765] 2.96375e-17        10   not_de ATAC_peak_263397  0.364738 2.08835e-08
##   [27766] 2.96375e-17        10   not_de ATAC_peak_263399  0.317387 1.20088e-08
##   -------
##   seqinfo: 25 sequences (1 circular) from hg38 genome</code></pre>
<p>Now we can ask, how many DA peaks are near DE genes relative to “other” non-DE genes? A gene may appear more than once in <code>genes_olap_peaks</code>, because multiple peaks may overlap a single gene, or because we have re-sampled the same gene more than once, or a combination of these two cases.</p>
<p>For each gene (that is the combination of chromosome, the start, end, and strand), and the “origin” (DE vs not-DE) we can compute the distinct number of peaks for each gene and the maximum peak based on LFC. This is achieved via <code>reduce_ranges_directed</code>, which allows an aggregation to result in a <em>GRanges</em> object via merging neighboring genomic regions. The use of the directed suffix indicates we’re maintaining strand information. In this case, we are simply merging ranges (genes) via the groups we mentioned above. We also have to account for the number of resamples we have performed when counting if there are any peaks, to ensure we do not double count the same peak:</p>
<div class="sourceCode" id="cb70"><html><body><pre class="r"><span class="no">gene_peak_max_lfc</span> <span class="kw">&lt;-</span> <span class="no">genes_olap_peaks</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">group_by</a></span>(<span class="no">gene_id</span>, <span class="no">origin</span>)  <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-reduce.html">reduce_ranges_directed</a></span>(
    <span class="kw">peak_count</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">da_padj</span>)) / <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n_distinct.html">n_distinct</a></span>(<span class="no">resample</span>),
    <span class="kw">peak_max_lfc</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">da_log2FC</span>))
  )</pre></body></html></div>
<p>We can then filter genes if they have any peaks and compare the peak fold changes between non-DE and DE genes using a boxplot:</p>
<div class="sourceCode" id="cb71"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="no">gene_peak_max_lfc</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">filter</a></span>(<span class="no">peak_count</span> <span class="kw">&gt;</span> <span class="fl">0</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">origin</span>, <span class="no">peak_max_lfc</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>()</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="fluentGenomics_files/figure-html/boxplot-1.png" alt="(ref:boxplot)" width="700"><p class="caption">
(ref:boxplot)
</p>
</div>
<p>In general, the DE genes have larger maximum DA fold changes relative to the non-DE genes.</p>
<p>Next we examine how thresholds on the DA LFC modify the enrichment we observe of DA peaks near DE or non-DE genes. First, we want to know how the number of peaks within DE genes and non-DE genes change as we change threshold values on the peak LFC. As an example, we could compute this by arbitrarily chosen LFC thresholds of 1 or 2 as follows:</p>
<div class="sourceCode" id="cb72"><html><body><pre class="r"><span class="no">origin_peak_lfc</span> <span class="kw">&lt;-</span> <span class="no">genes_olap_peaks</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">group_by</a></span>(<span class="no">origin</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">summarize</a></span>(
    <span class="kw">peak_count</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">da_padj</span>)) / <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n_distinct.html">n_distinct</a></span>(<span class="no">resample</span>),
    <span class="kw">lfc1_peak_count</span> <span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">da_log2FC</span>) <span class="kw">&gt;</span> <span class="fl">1</span>, <span class="kw">na.rm</span><span class="kw">=</span><span class="fl">TRUE</span>)/ <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n_distinct.html">n_distinct</a></span>(<span class="no">resample</span>),
    <span class="kw">lfc2_peak_count</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">da_log2FC</span>) <span class="kw">&gt;</span> <span class="fl">2</span>, <span class="kw">na.rm</span><span class="kw">=</span><span class="fl">TRUE</span>)/ <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n_distinct.html">n_distinct</a></span>(<span class="no">resample</span>)
  )
<span class="no">origin_peak_lfc</span></pre></body></html></div>
<pre><code>## DataFrame with 2 rows and 4 columns
##     origin peak_count lfc1_peak_count lfc2_peak_count
##   &lt;factor&gt;  &lt;numeric&gt;       &lt;numeric&gt;       &lt;numeric&gt;
## 1   not_de     2391.8           369.5            32.5
## 2   de         3416.0          1097.0           234.0</code></pre>
<p>Here we see that DE genes tend to have more DA peaks near them, and that the number of DA peaks decreases as we increase the DA LFC threshold (as expected). We now show how to compute the ratio of peak counts from DE compared to non-DE genes, so we can see how this ratio changes for various DA LFC thresholds.</p>
<p>For all variables except for the <code>origin</code> column we divide the first row’s values by the second row, which will be the enrichment of peaks in DE genes compared to other genes. This requires us to reshape the summary table from long form back to wide form using the <em>tidyr</em> package. First we pivot the results of the <code>peak_count</code> columns into name-value pairs, then pivot again to place values into the <code>origin</code> column. Then we create a new column with the relative enrichment:</p>
<div class="sourceCode" id="cb74"><html><body><pre class="r"><span class="no">origin_peak_lfc</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">tidyr</span><span class="kw ns">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span>(<span class="kw">cols</span> <span class="kw">=</span> -<span class="no">origin</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">tidyr</span><span class="kw ns">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span>(<span class="kw">names_from</span> <span class="kw">=</span> <span class="no">origin</span>, <span class="kw">values_from</span> <span class="kw">=</span> <span class="no">value</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">mutate</a></span>(<span class="kw">enrichment</span> <span class="kw">=</span> <span class="no">de</span> / <span class="no">not_de</span>)</pre></body></html></div>
<pre><code>## # A tibble: 3 x 4
##   name            not_de    de enrichment
##   &lt;chr&gt;            &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;
## 1 peak_count      2392.   3416       1.43
## 2 lfc1_peak_count  370.   1097       2.97
## 3 lfc2_peak_count   32.5   234       7.2</code></pre>
<p>The above table shows that relative enrichment increases for a larger LFC threshold.</p>
<p>Due to the one-to-many mappings of genes to peaks, it is unknown if we have the same number of DE genes participating or less, as we increase the threshold on the DA LFC. We can examine the number of genes with overlapping DA peaks at various thresholds by grouping and aggregating twice. First, the number of peaks that meet the thresholds are computed within each gene, origin, and resample group. Second, within the origin column, we compute the total number of peaks that meet the DA LFC threshold and the number of genes that have more than zero peaks (again averaging over the number of resamples).</p>
<div class="sourceCode" id="cb76"><html><body><pre class="r"><span class="no">genes_olap_peaks</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">group_by</a></span>(<span class="no">gene_id</span>, <span class="no">origin</span>, <span class="no">resample</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-reduce.html">reduce_ranges_directed</a></span>(
    <span class="kw">lfc1</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">da_log2FC</span>) <span class="kw">&gt;</span> <span class="fl">1</span>, <span class="kw">na.rm</span><span class="kw">=</span><span class="fl">TRUE</span>),
    <span class="kw">lfc2</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">da_log2FC</span>) <span class="kw">&gt;</span> <span class="fl">2</span>, <span class="kw">na.rm</span><span class="kw">=</span><span class="fl">TRUE</span>)
  ) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">group_by</a></span>(<span class="no">origin</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">summarize</a></span>(
    <span class="kw">lfc1_gene_count</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">lfc1</span> <span class="kw">&gt;</span> <span class="fl">0</span>) / <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n_distinct.html">n_distinct</a></span>(<span class="no">resample</span>),
    <span class="kw">lfc1_peak_count</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">lfc1</span>) / <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n_distinct.html">n_distinct</a></span>(<span class="no">resample</span>),
    <span class="kw">lfc2_gene_count</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">lfc2</span> <span class="kw">&gt;</span> <span class="fl">0</span>) / <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n_distinct.html">n_distinct</a></span>(<span class="no">resample</span>),
    <span class="kw">lfc2_peak_count</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">lfc2</span>) / <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n_distinct.html">n_distinct</a></span>(<span class="no">resample</span>)
  )</pre></body></html></div>
<pre><code>## DataFrame with 2 rows and 5 columns
##     origin lfc1_gene_count lfc1_peak_count lfc2_gene_count lfc2_peak_count
##   &lt;factor&gt;       &lt;numeric&gt;       &lt;numeric&gt;       &lt;numeric&gt;       &lt;numeric&gt;
## 1   not_de           271.2           369.5            30.3            32.5
## 2   de               515.0          1097.0           151.0           234.0</code></pre>
<p>To do this for many thresholds is cumbersome and would create a lot of duplicate code. Instead we create a single function called <code>count_above_threshold</code> that accepts a variable and a vector of thresholds, and computes the sum of the absolute value of the variable for each element in the <code>thresholds</code> vector.</p>
<div class="sourceCode" id="cb78"><html><body><pre class="r"><span class="no">count_if_above_threshold</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">var</span>, <span class="no">thresholds</span>) {
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">thresholds</span>, <span class="kw">function</span>(<span class="no">.</span>) <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">var</span>) <span class="kw">&gt;</span> <span class="no">.</span>, <span class="kw">na.rm</span> <span class="kw">=</span> <span class="fl">TRUE</span>))
}</pre></body></html></div>
<p>The above function will compute the counts for any arbitrary threshold, so we can apply it over possible LFC thresholds of interest. We choose a grid of one hundred thresholds based on the range of absolute LFC values in the <code>da_peaks</code> <em>GRanges</em> object:</p>
<div class="sourceCode" id="cb79"><html><body><pre class="r"><span class="no">thresholds</span> <span class="kw">&lt;-</span> <span class="no">da_peaks</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">mutate</a></span>(<span class="kw">abs_lfc</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">da_log2FC</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/with.html">with</a></span>(
    <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="no">abs_lfc</span>), <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">abs_lfc</span>), <span class="kw">length.out</span> <span class="kw">=</span> <span class="fl">100</span>)
  )</pre></body></html></div>
<p>The peak counts for each threshold are computed as a new list-column called <code>value</code>. First, the <em>GRanges</em> object has been grouped by the gene, origin, and the number of resamples columns. Then we aggregate over those columns, so each row will contain the peak counts for all of the thresholds for a gene, origin, and resample. We also maintain another list-column that contains the threshold values.</p>
<div class="sourceCode" id="cb80"><html><body><pre class="r"><span class="no">genes_peak_all_thresholds</span> <span class="kw">&lt;-</span> <span class="no">genes_olap_peaks</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">group_by</a></span>(<span class="no">gene_id</span>, <span class="no">origin</span>, <span class="no">resample</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-reduce.html">reduce_ranges_directed</a></span>(
    <span class="kw">value</span> <span class="kw">=</span> <span class="fu">count_if_above_threshold</span>(<span class="no">da_log2FC</span>, <span class="no">thresholds</span>),
    <span class="kw">threshold</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">thresholds</span>)
  )
<span class="no">genes_peak_all_thresholds</span></pre></body></html></div>
<pre><code>## GRanges object with 8239 ranges and 5 metadata columns:
##          seqnames              ranges strand |            gene_id   origin
##             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;character&gt; &lt;factor&gt;
##      [1]     chr1 196641878-196661877      + | ENSG00000000971.15       de
##      [2]     chr6   46119993-46139992      + |  ENSG00000001561.6       de
##      [3]     chr4   17567192-17587191      + | ENSG00000002549.12       de
##      [4]     chr7 150790403-150810402      + |  ENSG00000002933.8       de
##      [5]     chr4   15768275-15788274      + | ENSG00000004468.12       de
##      ...      ...                 ...    ... .                ...      ...
##   [8235]    chr17   43569620-43589619      - | ENSG00000175832.12   not_de
##   [8236]    chr17   18250534-18270533      + | ENSG00000177427.12   not_de
##   [8237]    chr20   63885182-63905181      + | ENSG00000101152.10   not_de
##   [8238]     chr1   39071316-39091315      + | ENSG00000127603.25   not_de
##   [8239]     chr8   41567187-41587186      + | ENSG00000158669.11   not_de
##           resample         value                         threshold
##          &lt;integer&gt; &lt;IntegerList&gt;                     &lt;NumericList&gt;
##      [1]         0     1,1,1,... 0.0658243,0.1184840,0.1711436,...
##      [2]         0     1,1,1,... 0.0658243,0.1184840,0.1711436,...
##      [3]         0     6,6,6,... 0.0658243,0.1184840,0.1711436,...
##      [4]         0     4,4,4,... 0.0658243,0.1184840,0.1711436,...
##      [5]         0  11,11,11,... 0.0658243,0.1184840,0.1711436,...
##      ...       ...           ...                               ...
##   [8235]        10     1,1,1,... 0.0658243,0.1184840,0.1711436,...
##   [8236]        10     3,3,2,... 0.0658243,0.1184840,0.1711436,...
##   [8237]        10     5,5,5,... 0.0658243,0.1184840,0.1711436,...
##   [8238]        10     3,3,3,... 0.0658243,0.1184840,0.1711436,...
##   [8239]        10     3,3,3,... 0.0658243,0.1184840,0.1711436,...
##   -------
##   seqinfo: 25 sequences (1 circular) from hg38 genome</code></pre>
<p>Now we can expand these list-columns into a long <em>GRanges</em> object using the <code><a href="https://rdrr.io/pkg/plyranges/man/ranges-expand.html">expand_ranges()</a></code> function. This function will unlist the <code>value</code> and <code>threshold</code> columns and lengthen the resulting <em>GRanges</em> object. To compute the peak and gene counts for each threshold, we apply the same summarization as before:</p>
<div class="sourceCode" id="cb82"><html><body><pre class="r"><span class="no">origin_peak_all_thresholds</span> <span class="kw">&lt;-</span> <span class="no">genes_peak_all_thresholds</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-expand.html">expand_ranges</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">group_by</a></span>(<span class="no">origin</span>, <span class="no">threshold</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">summarize</a></span>(
    <span class="kw">gene_count</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">value</span> <span class="kw">&gt;</span> <span class="fl">0</span>) / <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n_distinct.html">n_distinct</a></span>(<span class="no">resample</span>),
    <span class="kw">peak_count</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">value</span>) / <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n_distinct.html">n_distinct</a></span>(<span class="no">resample</span>)
  )
<span class="no">origin_peak_all_thresholds</span></pre></body></html></div>
<pre><code>## DataFrame with 200 rows and 4 columns
##       origin threshold gene_count peak_count
##     &lt;factor&gt; &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;
## 1     not_de 0.0658243      708.0     2391.4
## 2     not_de 0.1184840      698.8     2320.6
## 3     not_de 0.1711436      686.2     2178.6
## 4     not_de 0.2238033      672.4     1989.4
## 5     not_de 0.2764629      650.4     1785.8
## ...      ...       ...        ...        ...
## 196       de   5.06849          2          2
## 197       de   5.12115          0          0
## 198       de   5.17381          0          0
## 199       de   5.22647          0          0
## 200       de   5.27913          0          0</code></pre>
<p>Again we can compute the relative enrichment in LFCs in the same manner as before, by pivoting the results to long form then back to wide form to compute the enrichment. We visualize the peak enrichment changes of DE genes relative to other genes as a line chart:</p>
<div class="sourceCode" id="cb84"><html><body><pre class="r"><span class="no">origin_threshold_counts</span> <span class="kw">&lt;-</span> <span class="no">origin_peak_all_thresholds</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">tidyr</span><span class="kw ns">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span>(<span class="kw">cols</span> <span class="kw">=</span> -<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">origin</span>, <span class="no">threshold</span>),
                      <span class="kw">names_to</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"type"</span>, <span class="st">"var"</span>),
                      <span class="kw">names_sep</span> <span class="kw">=</span> <span class="st">"_"</span>,
                      <span class="kw">values_to</span> <span class="kw">=</span> <span class="st">"count"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">select</a></span>(-<span class="no">var</span>)

<span class="no">origin_threshold_counts</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">filter</a></span>(<span class="no">type</span> <span class="kw">==</span> <span class="st">"peak"</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">tidyr</span><span class="kw ns">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span>(<span class="kw">names_from</span> <span class="kw">=</span> <span class="no">origin</span>, <span class="kw">values_from</span> <span class="kw">=</span> <span class="no">count</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">mutate</a></span>(<span class="kw">enrichment</span> <span class="kw">=</span>  <span class="no">de</span> / <span class="no">not_de</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">threshold</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">enrichment</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="st">"logFC threshold"</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="st">"Relative Enrichment"</span>)</pre></body></html></div>
<pre><code>## Warning: Removed 4 row(s) containing missing values (geom_path).</code></pre>
<div class="figure" style="text-align: center">
<img src="fluentGenomics_files/figure-html/line-chart-1.png" alt="(ref:linechart)" width="700"><p class="caption">
(ref:linechart)
</p>
</div>
<p>We computed the sum of DA peaks near the DE genes, for increasing LFC thresholds on the accessibility change. As we increased the threshold, the number of total peaks went down (likewise the mean number of DA peaks per gene). It is also likely the number of DE genes with a DA peak nearby with such a large change went down. We can investigate this with a plot that summarizes many of the aspects underlying the enrichment plot above.</p>
<div class="sourceCode" id="cb86"><html><body><pre class="r"><span class="no">origin_threshold_counts</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">threshold</span>,
             <span class="kw">y</span> <span class="kw">=</span> <span class="no">count</span> + <span class="fl">1</span>,
             <span class="kw">color</span> <span class="kw">=</span> <span class="no">origin</span>,
             <span class="kw">linetype</span> <span class="kw">=</span> <span class="no">type</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>()</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="fluentGenomics_files/figure-html/line-chart2-1.png" alt="(ref:linechart2)" width="700"><p class="caption">
(ref:linechart2)
</p>
</div>
</div>
</div>
<div id="discussion" class="section level1">
<h1 class="hasAnchor">
<a href="#discussion" class="anchor"></a>Discussion</h1>
<p>We have shown that by using <em>plyranges</em> and <em>tximeta</em> (with support of Bioconductor and <em>tidyverse</em> ecosystems) we can fluently iterate through the biological data science workflow: from import, through to modeling, and data integration.</p>
<p>There are several further steps that would be interesting to perform in this analysis; for example, we could modify window size around the TSS to see how it affects enrichment, and vary the FDR cut-offs for both the DE gene and DA peak sets. We could also have computed variance in addition to the mean of the bootstrap set, and so drawn an interval around the enrichment line.</p>
<p>Finally, our workflow illustrates the benefits of using appropriate data abstractions provided by Bioconductor such as the <em>SummarizedExperiment</em> and <em>GRanges</em>. These abstractions provide users with a mental model of their experimental data and are the building blocks for constructing the modular and iterative analyses we have shown here. Consequently, we have been able to interoperate many decoupled R packages (from both Bioconductor and the tidyverse) to construct a seamless end-to-end workflow that is far too specialized for a single monolithic tool.</p>
</div>
<div id="software-availability" class="section level1">
<h1 class="hasAnchor">
<a href="#software-availability" class="anchor"></a>Software Availability</h1>
<p>The workflow materials, including this article can be fully reproduced following the instructions found at the Github repository <a href="https://github.com/sa-lee/fluentGenomics">sa-lee/fluentGenomics</a>. Moreover, the development version of the workflow and all downstream dependencies can be installed using the <code>BiocManager</code> package by running:</p>
<div class="sourceCode" id="cb87"><html><body><pre class="r"><span class="co"># development version from Github</span>
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"sa-lee/fluentGenomics"</span>)
<span class="co"># version available from Bioconductor</span>
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"fluentGenomics"</span>)</pre></body></html></div>
<p>This article and the analyses were performed with R <span class="citation">(R Core Team <a href="#ref-baser">2019</a>)</span> using the <em>rmarkdown</em> <span class="citation">(Allaire et al. <a href="#ref-rmarkdown">2019</a>)</span>, and <em>knitr</em> <span class="citation">(Xie <a href="#ref-knitr">2019</a>, <a href="#ref-xie2015">2015</a>)</span> packages.</p>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h2>
<div class="sourceCode" id="cb88"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R version 4.0.0 (2020-04-24)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.4 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] ggplot2_3.3.0               plyranges_1.9.0            
##  [3] DESeq2_1.29.4               GenomicFeatures_1.41.0     
##  [5] AnnotationDbi_1.51.0        SummarizedExperiment_1.19.4
##  [7] DelayedArray_0.15.1         matrixStats_0.56.0         
##  [9] Biobase_2.49.0              GenomicRanges_1.41.1       
## [11] GenomeInfoDb_1.25.0         IRanges_2.23.4             
## [13] S4Vectors_0.27.6            BiocGenerics_0.35.2        
## [15] readr_1.3.1                 dplyr_0.8.5                
## [17] tximeta_1.7.3               fluentGenomics_1.1.1       
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_1.4-1              ellipsis_0.3.1               
##   [3] rprojroot_1.3-2               XVector_0.29.0               
##   [5] fs_1.4.1                      farver_2.0.3                 
##   [7] bit64_0.9-7                   interactiveDisplayBase_1.27.0
##   [9] fansi_0.4.1                   splines_4.0.0                
##  [11] tximport_1.17.0               geneplotter_1.67.0           
##  [13] knitr_1.28                    jsonlite_1.6.1               
##  [15] Rsamtools_2.5.0               annotate_1.67.0              
##  [17] dbplyr_1.4.3                  shiny_1.4.0.2                
##  [19] BiocManager_1.30.10           compiler_4.0.0               
##  [21] httr_1.4.1                    backports_1.1.7              
##  [23] assertthat_0.2.1              Matrix_1.2-18                
##  [25] fastmap_1.0.1                 lazyeval_0.2.2               
##  [27] cli_2.0.2                     later_1.0.0                  
##  [29] htmltools_0.4.0               prettyunits_1.1.1            
##  [31] tools_4.0.0                   gtable_0.3.0                 
##  [33] glue_1.4.1                    GenomeInfoDbData_1.2.3       
##  [35] rappdirs_0.3.1                Rcpp_1.0.4.6                 
##  [37] pkgdown_1.5.1                 vctrs_0.3.0                  
##  [39] Biostrings_2.57.0             rtracklayer_1.49.1           
##  [41] xfun_0.13                     stringr_1.4.0                
##  [43] mime_0.9                      lifecycle_0.2.0              
##  [45] ensembldb_2.13.1              XML_3.99-0.3                 
##  [47] AnnotationHub_2.21.0          zlibbioc_1.35.0              
##  [49] MASS_7.3-51.5                 scales_1.1.1                 
##  [51] hms_0.5.3                     promises_1.1.0               
##  [53] ProtGenerics_1.21.0           AnnotationFilter_1.13.0      
##  [55] RColorBrewer_1.1-2            yaml_2.2.1                   
##  [57] curl_4.3                      memoise_1.1.0                
##  [59] biomaRt_2.45.0                stringi_1.4.6                
##  [61] RSQLite_2.2.0                 BiocVersion_3.12.0           
##  [63] highr_0.8                     genefilter_1.71.0            
##  [65] desc_1.2.0                    BiocParallel_1.23.0          
##  [67] rlang_0.4.6                   pkgconfig_2.0.3              
##  [69] bitops_1.0-6                  evaluate_0.14                
##  [71] lattice_0.20-41               purrr_0.3.4                  
##  [73] GenomicAlignments_1.25.0      labeling_0.3                 
##  [75] bit_1.1-15.2                  tidyselect_1.1.0             
##  [77] magrittr_1.5                  bookdown_0.19                
##  [79] R6_2.4.1                      DBI_1.1.0                    
##  [81] pillar_1.4.4                  withr_2.2.0                  
##  [83] survival_3.1-12               RCurl_1.98-1.2               
##  [85] tibble_3.0.1                  crayon_1.3.4                 
##  [87] utf8_1.1.4                    BiocFileCache_1.13.0         
##  [89] rmarkdown_2.1                 progress_1.2.2               
##  [91] locfit_1.5-9.4                grid_4.0.0                   
##  [93] blob_1.2.1                    digest_0.6.25                
##  [95] xtable_1.8-4                  tidyr_1.0.3                  
##  [97] httpuv_1.5.2                  openssl_1.4.1                
##  [99] munsell_0.5.0                 askpass_1.1</code></pre>
</div>
<div id="author-contributions" class="section level2">
<h2 class="hasAnchor">
<a href="#author-contributions" class="anchor"></a>Author Contributions</h2>
<p>All authors contributed to the writing and development of the workflow.</p>
</div>
<div id="competing-interests" class="section level2">
<h2 class="hasAnchor">
<a href="#competing-interests" class="anchor"></a>Competing interests</h2>
<p>The authors declare that they have no competing interests.</p>
</div>
<div id="funding" class="section level2">
<h2 class="hasAnchor">
<a href="#funding" class="anchor"></a>Funding</h2>
<p>SL is supported by an Australian Government Research Training Program (RTP) scholarship with a top up scholarship from CSL Limited.</p>
<p>MIL’s contribution is supported by NIH grant R01 HG009937.</p>
<p><strong>I confirm that the funders had no role in study design, data collection and analysis, decision to publish, or preparation of the manuscript.</strong></p>
</div>
<div id="acknowledgements" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h2>
<p>The authors would like to thank all participants of the Bioconductor 2019 and BiocAsia 2019 conferences who attended and provided feedback on early versions of this workflow paper.</p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-alasooZenodo">
<p>Alasoo, Kaur, and Daniel Gaffney. 2017. “Processed read counts from macrophage RNA-seq and ATAC-seq experiments.” Zenodo. <a href="https://doi.org/10.5281/zenodo.1188300">https://doi.org/10.5281/zenodo.1188300</a>.</p>
</div>
<div id="ref-alasoo">
<p>Alasoo, K, J Rodrigues, S Mukhopadhyay, AJ Knights, AL Mann, K Kundu, HIPSCI-Consortium, C Hale, Dougan G, and DJ Gaffney. 2018. “Shared genetic effects on chromatin and gene expression indicate a role for enhancer priming in immune response.” <em>Nature Genetics</em> 50: 424–31. <a href="https://doi.org/10.1038/s41588-018-0046-7">https://doi.org/10.1038/s41588-018-0046-7</a>.</p>
</div>
<div id="ref-rmarkdown">
<p>Allaire, JJ, Yihui Xie, Jonathan McPherson, Javier Luraschi, Kevin Ushey, Aron Atkins, Hadley Wickham, Joe Cheng, Winston Chang, and Richard Iannone. 2019. <em>Rmarkdown: Dynamic Documents for R</em>. <a href="https://github.com/rstudio/rmarkdown">https://github.com/rstudio/rmarkdown</a>.</p>
</div>
<div id="ref-gencode">
<p>Frankish, A, GENCODE-consoritum, and P Flicek. 2018. “GENCODE reference annotation for the human and mouse genomes.” <em>Nucleic Acids Research</em>. <a href="https://doi.org/10.1093/nar/gky955">https://doi.org/10.1093/nar/gky955</a>.</p>
</div>
<div id="ref-Hansen2012">
<p>Hansen, Kasper D., Rafael A. Irizarry, and Zhijin Wu. 2012. “Removing technical variability in RNA-seq data using condition quantile normalization.” <em>Biostatistics</em> 13 (2): 204–16. <a href="https://doi.org/10.1093/biostatistics/kxr054">https://doi.org/10.1093/biostatistics/kxr054</a>.</p>
</div>
<div id="ref-bioc">
<p>Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” <em>Nature Methods</em> 12 (2). Springer Nature: 115–21. <a href="https://doi.org/10.1038/nmeth.3252">https://doi.org/10.1038/nmeth.3252</a>.</p>
</div>
<div id="ref-snakemake">
<p>Köster, Johannes, and Sven Rahmann. 2012. “Snakemake—a scalable bioinformatics workflow engine.” <em>Bioinformatics</em> 28 (19): 2520–2. <a href="https://doi.org/10.1093/bioinformatics/bts480">https://doi.org/10.1093/bioinformatics/bts480</a>.</p>
</div>
<div id="ref-Law2018-f1000">
<p>Law, Charity W, Monther Alhamdoosh, Shian Su, Xueyi Dong, Luyi Tian, Gordon K Smyth, and Matthew E Ritchie. 2018. “RNA-seq Analysis Is Easy as 1-2-3 with Limma, Glimma and edgeR.” <em>F1000 Research</em> 5 (1408). F1000 Research Limited: 1408. <a href="https://doi.org/10.12688/f1000research.9005.3">https://doi.org/10.12688/f1000research.9005.3</a>.</p>
</div>
<div id="ref-granges">
<p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin Morgan, and Vincent Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput. Biol.</em> 9. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.</p>
</div>
<div id="ref-Lee2019">
<p>Lee, Stuart, Dianne Cook, and Michael Lawrence. 2019. “Plyranges: A Grammar of Genomic Data Transformation.” <em>Genome Biology</em> 20 (1): 4. <a href="https://doi.org/10.1186/s13059-018-1597-8">https://doi.org/10.1186/s13059-018-1597-8</a>.</p>
</div>
<div id="ref-Love2016-f1000">
<p>Love, Michael I, Simon Anders, Vladislav Kim, and Wolfgang Huber. 2016. “RNA-Seq Workflow: Gene-Level Exploratory Analysis and Differential Expression.” <em>F1000 Research</em> 4 (1070). F1000 Research Limited: 1070. <a href="https://doi.org/10.12688/f1000research.7035.2">https://doi.org/10.12688/f1000research.7035.2</a>.</p>
</div>
<div id="ref-Love2014">
<p>Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. “Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2.” <em>Genome Biology</em> 15 (12): 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.</p>
</div>
<div id="ref-Love2019-tximeta">
<p>Love, Michael I, Charlotte Soneson, Peter F Hickey, Lisa K Johnson, N Tessa Pierce, Lori Shepherd, Martin Morgan, and Rob Patro. 2019. “Tximeta: Reference Sequence Checksums for Provenance Identification in RNA-seq.” <em>bioRxiv</em>, September, 777888. <a href="https://doi.org/10.1101/777888">https://doi.org/10.1101/777888</a>.</p>
</div>
<div id="ref-salmon">
<p>Patro, R, G Duggal, MI Love, RA Irizarry, and C Kingsford. 2017. “Salmon Provides Fast and Bias-Aware Quantification of Transcript Expression.” <em>Nature Methods</em> 14: 417–19. <a href="https://doi.org/10.1038/nmeth.4197">https://doi.org/10.1038/nmeth.4197</a>.</p>
</div>
<div id="ref-baser">
<p>R Core Team. 2019. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.</p>
</div>
<div id="ref-bcfilecache">
<p>Shepherd, Lori, and Martin Morgan. 2019. <em>BiocFileCache: Manage Files Across Sessions</em>.</p>
</div>
<div id="ref-Smyth2004">
<p>Smyth, Gordon K. 2004. “Linear Models and Empirical Bayes Methods for Assessing Differential Expression in Microarray Experiments.” <em>Statistical Applications in Genetics and Molecular Biology</em> 3 (1).</p>
</div>
<div id="ref-Soneson2015">
<p>Soneson, Charlotte, Michael I. Love, and Mark Robinson. 2015. “Differential analyses for RNA-seq: transcript-level estimates improve gene-level inferences.” <em>F1000Research</em> 4 (1521). <a href="https://doi.org/10.12688/f1000research.7563.1">https://doi.org/10.12688/f1000research.7563.1</a>.</p>
</div>
<div id="ref-tidyverse">
<p>Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. “Welcome to the tidyverse.” <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.</p>
</div>
<div id="ref-xie2015">
<p>Xie, Yihui. 2015. <em>Dynamic Documents with R and Knitr</em>. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. <a href="https://yihui.name/knitr/">https://yihui.name/knitr/</a>.</p>
</div>
<div id="ref-knitr">
<p>———. 2019. <em>Knitr: A General-Purpose Package for Dynamic Report Generation in R</em>. <a href="https://yihui.name/knitr/">https://yihui.name/knitr/</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Stuart Lee, Michael Love.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
